File: app/src/main/java/com/dewonderstruck/apps/ashx0/viewmodel/product/ProductAttributeHeaderViewModel.java
package com.dewonderstruck.apps.ashx0.viewmodel.product;

import com.dewonderstruck.apps.ashx0.repository.product.ProductRepository;
import com.dewonderstruck.apps.ashx0.utils.AbsentLiveData;
import com.dewonderstruck.apps.ashx0.utils.Utils;
import com.dewonderstruck.apps.ashx0.viewmodel.common.PSViewModel;
import com.dewonderstruck.apps.ashx0.viewobject.ProductAttributeDetail;
import com.dewonderstruck.apps.ashx0.viewobject.ProductAttributeHeader;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.inject.Inject;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;

public class ProductAttributeHeaderViewModel extends PSViewModel {
    //for product attribute detail list

    private final LiveData<List<ProductAttributeHeader>> ProductAttributeHeaderListData;
    private MutableLiveData<ProductAttributeHeaderViewModel.TmpDataHolder> ProductAttributeHeaderObj = new MutableLiveData<>();

    public String headerId;
    public boolean isHeaderData = false;
    public ProductAttributeDetail productAttributeDetail;
    public float price = 0;
    public float originalPrice = 0;

    public List<String> headerIdList = new ArrayList<>();

    public Map<String, String> basketItemHolderHashMap = new HashMap<>();
    public Map<String, Integer> attributeHeaderHashMap = new HashMap<>();
    public float priceAfterAttribute = 0;
    public float originalPriceAfterAttribute = 0;

    //endregion

    //region Constructor

    @Inject
    public ProductAttributeHeaderViewModel(ProductRepository productRepository) {
        //  product attribute detail List
        ProductAttributeHeaderListData = Transformations.switchMap(ProductAttributeHeaderObj, (ProductAttributeHeaderViewModel.TmpDataHolder obj) -> {
            if (obj == null) {
                return AbsentLiveData.create();
            }
            Utils.psLog("ProductAttributeHeaderListData");
            return productRepository.getProductAttributeHeader(obj.productId);
        });

    }

    //endregion
    //region Getter And Setter for product attribute detail List

    public void setProductAttributeHeaderListObj(String productId) {

        ProductAttributeHeaderViewModel.TmpDataHolder tmpDataHolder = new ProductAttributeHeaderViewModel.TmpDataHolder();
        tmpDataHolder.productId = productId;
        ProductAttributeHeaderObj.setValue(tmpDataHolder);

    }

    public LiveData<List<ProductAttributeHeader>> getProductAttributeHeaderListData() {
        return ProductAttributeHeaderListData;
    }

    //endregion

    //region Holder
    static class TmpDataHolder {
        public String offset = "";
        public String productId = "";
        public Boolean isConnected = false;
    }
    //endregion
}

File: app/src/main/java/com/dewonderstruck/apps/ashx0/viewmodel/product/ProductAttributeHeaderViewModel.kt
package com.dewonderstruck.apps.ashx0.viewmodel.product

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import com.dewonderstruck.apps.ashx0.repository.product.ProductRepository
import com.dewonderstruck.apps.ashx0.utils.AbsentLiveData
import com.dewonderstruck.apps.ashx0.utils.Utils
import com.dewonderstruck.apps.ashx0.viewmodel.common.PSViewModel
import com.dewonderstruck.apps.ashx0.viewobject.ProductAttributeDetail
import com.dewonderstruck.apps.ashx0.viewobject.ProductAttributeHeader
import java.util.*
import javax.inject.Inject

class ProductAttributeHeaderViewModel @Inject constructor(productRepository: ProductRepository) : PSViewModel() {
    //for product attribute detail list
    val productAttributeHeaderListData: LiveData<List<ProductAttributeHeader>>
    private val ProductAttributeHeaderObj = MutableLiveData<TmpDataHolder>()
    var headerId: String? = null
    @JvmField
    var isHeaderData = false
    @JvmField
    var productAttributeDetail: ProductAttributeDetail? = null
    @JvmField
    var price = 0f
    @JvmField
    var originalPrice = 0f
    @JvmField
    var headerIdList: List<String> = ArrayList()
    @JvmField
    var basketItemHolderHashMap: Map<String, String> = HashMap()
    @JvmField
    var attributeHeaderHashMap: Map<String, Int> = HashMap()
    @JvmField
    var priceAfterAttribute = 0f
    @JvmField
    var originalPriceAfterAttribute = 0f

    //endregion
    //region Getter And Setter for product attribute detail List
    fun setProductAttributeHeaderListObj(productId: String) {
        val tmpDataHolder = TmpDataHolder()
        tmpDataHolder.productId = productId
        ProductAttributeHeaderObj.value = tmpDataHolder
    }

    //endregion
    //region Holder
    internal class TmpDataHolder {
        var offset = ""
        var productId = ""
        var isConnected = false
    } //endregion

    //endregion
    //region Constructor
    init {
        //  product attribute detail List
        productAttributeHeaderListData = Transformations.switchMap(ProductAttributeHeaderObj) { obj: TmpDataHolder? ->
            if (obj == null) {
                return@switchMap AbsentLiveData.create<List<ProductAttributeHeader>>()
            }
            Utils.psLog("ProductAttributeHeaderListData")
            productRepository.getProductAttributeHeader(obj.productId)
        }
    }
}